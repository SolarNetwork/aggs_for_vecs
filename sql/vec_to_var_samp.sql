-- smallint var_samp lots
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- smallint var_samp none
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id = -1;
-- smallint var_samp one null
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id = 1;
-- smallint var_samp array of nulls
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id = 2;
-- smallint var_samp one not-null
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id = 4;
-- smallint var_samp array of nulls and one other
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id IN (2, 4);

-- int var_samp lots
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- int var_samp none
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id = -1;
-- int var_samp one null
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id = 1;
-- int var_samp array of nulls
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id = 2;
-- int var_samp one not-null
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id = 4;
-- int var_samp array of nulls and one other
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id IN (2, 4);

-- bigint var_samp lots
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- bigint var_samp none
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id = -1;
-- bigint var_samp one null
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id = 1;
-- bigint var_samp array of nulls
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id = 2;
-- bigint var_samp one not-null
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id = 4;
-- bigint var_samp array of nulls and one other
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id IN (2, 4);

-- real var_samp lots
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- real var_samp none
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id = -1;
-- real var_samp one null
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id = 1;
-- real var_samp array of nulls
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id = 2;
-- real var_samp one not-null
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id = 4;
-- real var_samp array of nulls and one other
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id IN (2, 4);

-- float8 var_samp lots
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- float8 var_samp none
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = -1;
-- float8 var_samp one null
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = 1;
-- float8 var_samp array of nulls
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = 2;
-- float8 var_samp one not-null
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = 4;
-- float8 var_samp array of nulls and one other
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id IN (2, 4);

-- numeric var_samp lots
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- numeric var_samp none
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = -1;
-- numeric var_samp one null
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = 1;
-- numeric var_samp array of nulls
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = 2;
-- numeric var_samp one not-null
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = 4;
-- numeric var_samp array of nulls and one other
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id IN (2, 4);
-- numeric var_samp data 01 data_i
-- results expected to match:
-- WITH di AS (
-- 	SELECT idx_i, var_samp(val_i) AS var_samp_i
-- 	FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
-- 	GROUP BY idx_i
-- )
-- SELECT array_agg(var_samp_i ORDER BY idx_i) AS data_i_var_samp FROM di
SELECT vec_to_var_samp(data_i) FROM measurements2;
-- numeric var_samp data 01 data_a
-- results expected to match:
-- WITH da AS (
-- 	SELECT idx_a, var_samp(val_a) AS var_samp_a
-- 	FROM measurements2 d, unnest(d.data_a) WITH ORDINALITY AS a(val_a, idx_a)
-- 	GROUP BY idx_a
-- )
-- SELECT array_agg(var_samp_a ORDER BY idx_a) AS data_a_var_samp FROM da
SELECT vec_to_var_samp(pad_vec(data_a, 3)) FROM measurements2;
