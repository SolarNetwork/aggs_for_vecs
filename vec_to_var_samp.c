
Datum vec_to_var_samp_transfn(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_to_var_samp_transfn);

/**
 * Returns an of n elements,
 * which each element is the sample variance value found in that position
 * from all input arrays.
 *
 * We use the sample approach as the built in VAR_SAMP aggregate.
 * See float8_var_samp in backend/utils/adt/float.c for details.
 * Our transition function records N, sum(X), and sum(X^2) for each array element,
 * and then the final function uses those to compute the sample variance,
 * based on:
 *
 *     s^2 = \frac{ \sum{}X^2 - \frac{(\sum{}X)^2}{N} }{N - 1}
 *
 * aka
 *           Σ X^2 - ((Σ X)^2)/N
 *     s^2 = -------------------
 *                  N - 1
 *
 * Using this formula rather than the more common one
 * lets us keep just 3 numbers (for each array position) as we go,
 * whereas to find X - mean(X) for each X
 * we'd have to keep all the values around until we knew the mean.
 *
 * This can over/underflow, but we report an error if that happens,
 * and it's no worse than the built-in aggregate.
 *
 * by Paul A. Jungwirth
 */
Datum
vec_to_var_samp_transfn(PG_FUNCTION_ARGS)
{
  Oid elemTypeId;
  int16 elemTypeWidth;
  bool elemTypeByValue;
  char elemTypeAlignmentCode;
  int currentLength;
  MemoryContext aggContext;
  VecArrayBuildState *state = NULL;
  ArrayType *currentArray;
  int arrayLength;
  Datum *currentVals;
  bool *currentNulls;
  int i;
  float tmp_f;

  if (!AggCheckCallContext(fcinfo, &aggContext)) {
    elog(ERROR, "vec_to_var_samp_transfn called in non-aggregate context");
  }

  // PG_ARGISNULL tests for SQL NULL,
  // but after the first pass we can have a
  // value that is non-SQL-NULL but still is C NULL.
  if (!PG_ARGISNULL(0)) {
    state = (VecArrayBuildState *)PG_GETARG_POINTER(0);
  }

  if (PG_ARGISNULL(1)) {
    // just return the current state unchanged (possibly still NULL)
    PG_RETURN_POINTER(state);
  }
  currentArray = PG_GETARG_ARRAYTYPE_P(1);

  if (state == NULL) {
    // Since we have our first not-null argument
    // we can initialize the state to match its length.
    elemTypeId = ARR_ELEMTYPE(currentArray);
    if (elemTypeId != INT2OID &&
        elemTypeId != INT4OID &&
        elemTypeId != INT8OID &&
        elemTypeId != FLOAT4OID &&
        elemTypeId != FLOAT8OID) {
      ereport(ERROR, (errmsg("vec_to_var_samp input must be array of SMALLINT, INTEGER, BIGINT, REAL, or DOUBLE PRECISION")));
    }
    if (ARR_NDIM(currentArray) != 1) {
      ereport(ERROR, (errmsg("One-dimensional arrays are required")));
    }
    arrayLength = (ARR_DIMS(currentArray))[0];
    // Just start with all NULLs and let the comparisons below replace them:
    state = initVecArrayResultWithNulls(elemTypeId, FLOAT8OID, aggContext, arrayLength);
  } else {
    elemTypeId = state->inputElementType;
    arrayLength = state->state.nelems;
  }

  get_typlenbyvalalign(elemTypeId, &elemTypeWidth, &elemTypeByValue, &elemTypeAlignmentCode);
  deconstruct_array(currentArray, elemTypeId, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode,
      &currentVals, &currentNulls, &currentLength);
  if (currentLength != arrayLength) {
    ereport(ERROR, (errmsg("All arrays must be the same length, but we got %d vs %d", currentLength, arrayLength)));
  }

  for (i = 0; i < arrayLength; i++) {
    if (currentNulls[i]) {
      // do nothing: nulls can't change the result.
    } else if (state->veccounts[i] == 0) {
      state->veccounts[i] = 1;
      switch (elemTypeId) {
        case INT2OID:   state->vecvalues[i].f8 = DatumGetInt16(currentVals[i]);  break;
        case INT4OID:   state->vecvalues[i].f8 = DatumGetInt32(currentVals[i]);  break;
        case INT8OID:   state->vecvalues[i].f8 = DatumGetInt64(currentVals[i]);  break;
        case FLOAT4OID: state->vecvalues[i].f8 = DatumGetFloat4(currentVals[i]); break;
        case FLOAT8OID: state->vecvalues[i].f8 = DatumGetFloat8(currentVals[i]); break;
        default: elog(ERROR, "Unknown elemTypeId!");
      }
      state->vectmpvalues[i].f8 = state->vecvalues[i].f8 * state->vecvalues[i].f8;
    } else {
      state->veccounts[i] += 1;
      switch (elemTypeId) {
        case INT2OID:   tmp_f = DatumGetInt16(currentVals[i]);  break;
        case INT4OID:   tmp_f = DatumGetInt32(currentVals[i]);  break;
        case INT8OID:   tmp_f = DatumGetInt64(currentVals[i]);  break;
        case FLOAT4OID: tmp_f = DatumGetFloat4(currentVals[i]); break;
        case FLOAT8OID: tmp_f = DatumGetFloat8(currentVals[i]); break;
        default: elog(ERROR, "Unknown elemTypeId!");
      }
      state->vecvalues[i].f8    += tmp_f;
      state->vectmpvalues[i].f8 += tmp_f * tmp_f;
    }
  }
  PG_RETURN_POINTER(state);
}

Datum vec_to_var_samp_finalfn(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_to_var_samp_finalfn);

Datum
vec_to_var_samp_finalfn(PG_FUNCTION_ARGS)
{
  Datum result;
  VecArrayBuildState *state;
  int dims[1];
  int lbs[1];
  int i;
  float8 numerator;

  Assert(AggCheckCallContext(fcinfo, NULL));

  state = PG_ARGISNULL(0) ? NULL : (VecArrayBuildState *)PG_GETARG_POINTER(0);

  if (state == NULL) PG_RETURN_NULL();

  // Convert from our pgnums to Datums:
  for (i = 0; i <= state->state.nelems; i++) {
    // Sample variance is undefined if N is 0 *or* 1:
    if (state->veccounts[i] > 1) {
      state->state.dnulls[i] = false;

      numerator = state->veccounts[i] * state->vectmpvalues[i].f8 - state->vecvalues[i].f8 * state->vecvalues[i].f8;
      CHECKFLOATVAL(numerator, isinf(state->vectmpvalues[i].f8) || isinf(state->vecvalues[i].f8), true);

      /* Watch out for roundoff error producing a negative numerator */
      if (numerator <= 0.0) {
        state->state.dvalues[i] = Float8GetDatum(0.0);
      } else {
        state->state.dvalues[i] = Float8GetDatum(numerator / (state->veccounts[i] * (state->veccounts[i] - 1)));
      }
    }
  }

  dims[0] = state->state.nelems;
  lbs[0] = 1;

  result = makeMdArrayResult(&state->state, 1, dims, lbs, CurrentMemoryContext, false);
  PG_RETURN_DATUM(result);
}
