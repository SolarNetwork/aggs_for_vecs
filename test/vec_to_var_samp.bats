load test_helper

@test "int16 var_samp" {
  result="$(query "SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{0,NULL,0.5}" ]
}

# @test "int32 var_samp" {
  # result="$(query "SELECT vec_to_var_samp('{1,1,5,2,0}'::integer[])")";
  # [ "$result" = "0" ]
# }

# @test "int64 var_samp" {
  # result="$(query "SELECT vec_to_var_samp('{1,1,5,2,0}'::bigint[])")";
  # [ "$result" = "0" ]
# }

# @test "float4 var_samp" {
  # result="$(query "SELECT vec_to_var_samp('{1,1,5,2,0}'::real[])")";
  # [ "$result" = "0" ]
# }

@test "float8 var_samp lots" {
  result="$(query "SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{0,NULL,0.5}" ]
}

@test "float8 var_samp none" {
  result="$(query "SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 var_samp one null" {
  result="$(query "SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 var_samp array of nulls" {
  result="$(query "SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "float8 var_samp one not-null" {
  result="$(query "SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "float8 var_samp array of nulls and one other" {
  result="$(query "SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "numeric var_samp lots" {
  result="$(query "SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{0.00000000000000000000,NULL,0.61605000000000000000}" ]
}

@test "numeric var_samp none" {
  result="$(query "SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "numeric var_samp one null" {
  result="$(query "SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "numeric var_samp array of nulls" {
  result="$(query "SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "numeric var_samp one not-null" {
  result="$(query "SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "numeric var_samp array of nulls and one other" {
  result="$(query "SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

# "numeric var_samp data 01 data_i" results calculated via:
# WITH di AS (
# 	SELECT idx_i, var_samp(val_i) AS var_samp_i
# 	FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
# 	GROUP BY idx_i
# )
# SELECT array_agg(var_samp_i ORDER BY idx_i) AS data_i_var_samp FROM di
#
# Giving _very slightly_ different results from vec_to_var_samp; have not figured out why, but
# assuming difference in scale handling for var_samp vs manual numeric calculations. Above gives:
# {3010518348472.5286,45266160.727322130532,12.8248416278119149,0.13484388848692884296,1007125661019.31762295,0.00445150048159007602,49464.85046351161695,4016221414458.0807,3036555176748.7055}

@test "numeric var_samp data 01 data_i" {
  result="$(query "SELECT vec_to_var_samp(data_i) FROM measurements2")";
  echo $result;
  [ "$result" = "{3010518348472.5286,45266160.727322130532,12.8248416278119126,0.13484388848692896175,1007125661019.3176,0.00445150048159007596,49464.8504635116169504,4016221414458.0807,3036555176748.7055}" ]
}

# "numeric var_samp data 01 data_a" results calculated via:
# WITH da AS (
# 	SELECT idx_a, var_samp(val_a) AS var_samp_a
# 	FROM measurements2 d, unnest(d.data_a) WITH ORDINALITY AS a(val_a, idx_a)
# 	GROUP BY idx_a
# )
# SELECT array_agg(var_samp_a ORDER BY idx_a) AS data_a_var_samp FROM da
#
# Giving _very slightly_ different results from vec_to_var_samp; have not figured out why, but
# assuming difference in scale handling for var_samp vs manual numeric calculations. Above gives:
# {3111992499.95126515,NULL,0}

@test "numeric var_samp data 01 data_a" {
  result="$(query "SELECT vec_to_var_samp(pad_vec(data_a, 3)) FROM measurements2")";
  echo $result;
  [ "$result" = "{3111992499.95126503,NULL,0.00000000000000000000}" ]
}
