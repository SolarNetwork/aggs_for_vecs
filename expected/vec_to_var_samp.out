-- smallint var_samp lots
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_var_samp 
-----------------
 {0,NULL,0.5}
(1 row)

-- smallint var_samp none
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id = -1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- smallint var_samp one null
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id = 1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- smallint var_samp array of nulls
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id = 2;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- smallint var_samp one not-null
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id = 4;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- smallint var_samp array of nulls and one other
SELECT vec_to_var_samp(smallints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- int var_samp lots
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_var_samp 
-----------------
 {0,NULL,0.5}
(1 row)

-- int var_samp none
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id = -1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- int var_samp one null
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id = 1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- int var_samp array of nulls
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id = 2;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- int var_samp one not-null
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id = 4;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- int var_samp array of nulls and one other
SELECT vec_to_var_samp(ints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- bigint var_samp lots
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_var_samp 
-----------------
 {0,NULL,0.5}
(1 row)

-- bigint var_samp none
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id = -1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- bigint var_samp one null
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id = 1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- bigint var_samp array of nulls
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id = 2;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- bigint var_samp one not-null
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id = 4;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- bigint var_samp array of nulls and one other
SELECT vec_to_var_samp(bigints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- real var_samp lots
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_var_samp 
-----------------
 {0,NULL,0.5}
(1 row)

-- real var_samp none
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id = -1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- real var_samp one null
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id = 1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- real var_samp array of nulls
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id = 2;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- real var_samp one not-null
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id = 4;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- real var_samp array of nulls and one other
SELECT vec_to_var_samp(reals) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- float8 var_samp lots
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_var_samp 
-----------------
 {0,NULL,0.5}
(1 row)

-- float8 var_samp none
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = -1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- float8 var_samp one null
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = 1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- float8 var_samp array of nulls
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = 2;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- float8 var_samp one not-null
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id = 4;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- float8 var_samp array of nulls and one other
SELECT vec_to_var_samp(floats) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- numeric var_samp lots
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
                   vec_to_var_samp                    
------------------------------------------------------
 {0.00000000000000000000,NULL,0.61605000000000000000}
(1 row)

-- numeric var_samp none
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = -1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- numeric var_samp one null
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = 1;
 vec_to_var_samp 
-----------------
 
(1 row)

-- numeric var_samp array of nulls
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = 2;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- numeric var_samp one not-null
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id = 4;
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- numeric var_samp array of nulls and one other
SELECT vec_to_var_samp(nums) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_var_samp  
------------------
 {NULL,NULL,NULL}
(1 row)

-- numeric var_samp data 01 data_i
-- results expected to match:
-- WITH di AS (
-- 	SELECT idx_i, var_samp(val_i) AS var_samp_i
-- 	FROM measurements2 d, unnest(d.data_i) WITH ORDINALITY AS a(val_i, idx_i)
-- 	GROUP BY idx_i
-- )
-- SELECT array_agg(var_samp_i ORDER BY idx_i) AS data_i_var_samp FROM di
SELECT vec_to_var_samp(data_i) FROM measurements2;
                                                                                        vec_to_var_samp                                                                                         
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {3010518348472.5286,45266160.727322130532,12.8248416278119149,0.13484388848692884296,1007125661019.31762295,0.00445150048159007602,49464.85046351161695,4016221414458.0807,3036555176748.7055}
(1 row)

-- numeric var_samp data 01 data_a
-- results expected to match:
-- WITH da AS (
-- 	SELECT idx_a, var_samp(val_a) AS var_samp_a
-- 	FROM measurements2 d, unnest(d.data_a) WITH ORDINALITY AS a(val_a, idx_a)
-- 	GROUP BY idx_a
-- )
-- SELECT array_agg(var_samp_a ORDER BY idx_a) AS data_a_var_samp FROM da
SELECT vec_to_var_samp(pad_vec(data_a, 3)) FROM measurements2;
                    vec_to_var_samp                    
-------------------------------------------------------
 {3111992499.95126515,NULL,0.000000000000000000000000}
(1 row)

